config/resolver_test.go

```diff
@@ -29,3 +31,101 @@ func TestValidatorImpl_VerifyShouldSetDefaultValue(t *testing.T) {
 	_ = v.Verify()
 	assert.Equal(t, "set", mc.Foo.FieldToSet)
 }
+
+func TestResolveImpl_WatchConfigShouldCallCallbackOnlyOnConfigurationContentChange(t *testing.T) {
+	type Config struct {
+		Field1 string `yaml:"field1"`
+	}
+
+	const configFile = "ut_resolve_1.yaml"
+	const initialContent = "field1: toto"
+	const changedContent = "field1: yoyo"
+
+	err := os.WriteFile(configFile, []byte(initialContent), 0777)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer os.Remove(configFile)
+
+	var config Config
+
+	callbackCallCount := 0
+	err = NewResolver[Config]().
+		SetConfigFile(configFile).
+		AddChangeCallback(func(newConfig *Config) {
+			callbackCallCount++
+		}).
+		Resolve(&config).
+		Verify()
+
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	// No change, callbacks shouldnt be called
+	err = os.WriteFile(configFile, []byte(initialContent), 0777)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	time.Sleep(50 * time.Millisecond)
+
+	assert.Equal(t, 0, callbackCallCount)
+
+	// Changes done, callbacks must be called
+	err = os.WriteFile(configFile, []byte(changedContent), 0777)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	time.Sleep(50 * time.Millisecond)
+
+	assert.Equal(t, 1, callbackCallCount)
+	assert.Equal(t, "yoyo", config.Field1)
+}
+
+func TestResolveImpl_WatchSliceConfigShouldApplyChanges(t *testing.T) {
+	type Config []int
+
+	const configFile = "ut_resolve_2.yaml"
+	const initialContent = "[0,1]"
+	const changedContent = "[3,4,5]"
+
+	err := os.WriteFile(configFile, []byte(initialContent), 0777)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer os.Remove(configFile)
+
+	var config Config
+
+	callbackCallCount := 0
+	err = NewResolver[Config]().
+		SetConfigFile(configFile).
+		AddChangeCallback(func(newConfig *Config) {
+			callbackCallCount++
+		}).
+		Resolve(&config).
+		Verify()
+
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	// Changes done, callbacks must be called
+	err = os.WriteFile(configFile, []byte(changedContent), 0777)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	time.Sleep(50 * time.Millisecond)
+
+	assert.Equal(t, 1, callbackCallCount)
+	assert.Equal(t, 3, config[0])
```

Comment:
I understand that the variable passed to Resolve is also changed by the watcher ? 

I'm not sure that it is a good idea. Having a callback registration let suppose that the variable is not changed, otherwise why have a callback ?

Reply:
I'm not sure you will be able to do the necessary new(T) in your code to make sure you don't reuse the original object. To be checked

Reply:
Initially i used a new pointer to the config but i reverted it to use the config pointer given in the Resolve() func
As the user gives a pointer to the config, i think we should use the same config pointer for the callback

Reply:
Having the configuration object changed automatically is still not a good idea to me. I prefer to rely only on a callback or a separate channel. So that it is easier to manage eventual concurrency access to the configuration or recreate objects depending on the config if needed.

Reply:
ok then i'll change this